language-support/ts/codegen/src/Main.hs

```diff
@@ -59,61 +58,61 @@ daml2ts Options{..} pkgId pkg = do
     T.writeFileUtf8 (outputDir </> "packageId.ts") $ T.unlines
         ["export default '" <> unPackageId pkgId <> "';"]
     forM_ (packageModules pkg) $ \mod -> do
-        let outputFile = outputDir </> joinPath (map T.unpack (unModuleName (moduleName mod))) <.> "ts"
-        putStrLn $ "Generating " ++ outputFile
-        createDirectoryIfMissing True (takeDirectory outputFile)
-        T.writeFileUtf8 outputFile (genModule mod)
+        whenJust (genModule mod) $ \modTxt -> do
+            let outputFile = outputDir </> joinPath (map T.unpack (unModuleName (moduleName mod))) <.> "ts"
+            putStrLn $ "Generating " ++ outputFile
+            createDirectoryIfMissing True (takeDirectory outputFile)
+            T.writeFileUtf8 outputFile modTxt
 
 dup :: a -> (a, a)
 dup x = (x, x)
 
-genModule :: Module -> T.Text
-genModule mod =
+genModule :: Module -> Maybe T.Text
+genModule mod
+  | null serDefs = Nothing
+  | otherwise =
     let curModName = moduleName mod
         pkgRootPath
           | lenModName == 1 = "."
           | otherwise = T.intercalate "/" (replicate (lenModName - 1) "..")
           where
             lenModName = length (unModuleName curModName)
         tpls = moduleTemplates mod
+        (defSers, refs) = unzip (map (genDefDataType curModName tpls) serDefs)
     in
-    T.unlines $
+    Just $ T.unlines $
         ["// Generated from " <> T.intercalate "/" (unModuleName curModName) <> ".daml"
         ,"/* eslint-disable @typescript-eslint/camelcase */"
-        ,"/* eslint-disable @typescript-eslint/no-unused-vars */"
         ,"/* eslint-disable @typescript-eslint/no-use-before-define */"
         ,"import * as daml from '@digitalasset/daml-json-types';"
         ,"import * as jtv from '@mojotech/json-type-validation';"
         ,"import packageId from '" <> pkgRootPath <> "/packageId';"
         ] ++
         ["import * as " <> modNameStr <> " from '" <> pkgRootPath <> "/" <> pkgRefStr <> T.intercalate "/" (unModuleName modName) <> "';"
-        | modRef@(pkgRef, modName) <- Set.toList (Set.setOf moduleModuleRef mod)
+        | modRef@(pkgRef, modName) <- Set.toList (Set.unions refs)
         , let pkgRefStr = case pkgRef of
                 PRSelf -> ""
                 PRImport pkgId -> "../" <> unPackageId pkgId <> "/"
-        , Just modNameStr <- [genModuleRef curModName modRef]
+        , let modNameStr = genModuleRef modRef
         ] ++
         [ ""
         ,"const moduleName = '" <> T.intercalate "." (unModuleName curModName) <> "';"
         ,"const templateId = (entityName: string): daml.TemplateId => ({packageId, moduleName, entityName});"
         ] ++
-        concat
-        [ [""] ++ def' ++ ser
-        | def <- NM.toList (moduleDataTypes mod)
-        , getIsSerializable (dataSerializable def)
-        , let (def', ser) = genDefDataType curModName tpls def
-        ]
+        concatMap (\(def, ser) -> [""] ++ def ++ ser) defSers
```

Comment:
imhoo this would be easier to read as something like
```
++ [""] ++
intercalate [""] $ concatMap (\(def, ser) -> def ++ ser) defSers
```
It took me a while to spot why we add a list with an empty string to each definition.